storage/remote/write_handler.go

```diff
@@ -479,6 +499,42 @@ func (h *writeHandler) appendV2(app storage.Appender, req *writev2.Request, rs *
 	return samplesWithoutMetadata, http.StatusBadRequest, errors.Join(badRequestErrs...)
 }
 
+// handleCTZeroSample appends CT as a zero-value sample with CT value as the sample timestamp.
+// It doens't return errors in case of out of order CT.
+func (h *writeHandler) handleCTZeroSample(app storage.Appender, ref storage.SeriesRef, l labels.Labels, sample writev2.Sample, ct int64) (storage.SeriesRef, error) {
+	var err error
+	if sample.Timestamp != 0 && ct != 0 {
```

Comment:
My preference would be to:

* Create `shouldIngestCTZeroSample() bool ` with `h.ingestCTZeroSample && ct != 0 && len(ts.Samples) > 0 && sample.Timestamp != 0` (notice you put more likely things to AND if first).
* Merge or inline both histogram and sample implementation as the err handling is exactly the same.

The main reason is the confusion on how to use this shallow function:

<img width="839" alt="image" src="https://github.com/user-attachments/assets/a09f9a0f-6cf1-428d-aa82-3ea0a8ccc24f">

I looked at it and I was like "AHA, there's bug! CT or ts might be missing!" and then I checked function etc. The main problem is that you check some stuff related to CT on caller side, some on implementation side. I think it's better to do all on caller side (or essentially in the same place).

Reply:
Just to clarify, do you prefer that:

1) We have both `shouldAppendCTZeroSample()` and `handleCTZeroSample` while moving all conditionals to `shouldAppendCTZeroSample()` and `handleCTZeroSample` would just call appendCT and rewrite the error if it is OOO.

2) We only have `shouldAppendCTZeroSample()` with all the conditionals and what's left from `handleCTZeroSample` is moved upwards to `appendv2`

?

Reply:
Sounds like it's simple enough to do should bool fn + inline

Reply:
While working on this I realized that shouldAppendCTZeroSample() would be a shallow function, so I just inlined this part as well 😬